fd/jacobian.go

```diff
@@ -0,0 +1,204 @@
+	}
+	r, c := dst.Dims()
+	if r != m || c != n {
+		panic("jacobian: mismatched matrix size")
+	}
+
+	if settings == nil {
+		settings = &JacobianSettings{}
+	}
+	if settings.OriginValue != nil && len(settings.OriginValue) != m {
+		panic("jacobian: mismatched OriginValue slice length")
+	}
+
+	formula := settings.Formula
+	if formula.isZero() {
+		formula = Forward
+	}
+	if formula.Derivative == 0 || formula.Stencil == nil || formula.Step == 0 {
+		panic("jacobian: bad formula")
+	}
+	if formula.Derivative != 1 {
+		panic("jacobian: invalid derivative order")
+	}
+
+	step := settings.Step
+	if step == 0 {
+		step = formula.Step
+	}
+
+	evals := n * len(formula.Stencil)
+	for _, pt := range formula.Stencil {
+		if pt.Loc == 0 {
+			evals -= n - 1
+			break
+		}
+	}
+	nWorkers := 1
+	if settings.Concurrent {
+		nWorkers = runtime.GOMAXPROCS(0)
+		if nWorkers > evals {
+			nWorkers = evals
+		}
+	}
+	if nWorkers == 1 {
+		jacobianSerial(dst, f, x, settings.OriginValue, formula, step)
+	} else {
+		jacobianConcurrent(dst, f, x, settings.OriginValue, formula, step, nWorkers)
+	}
+
+	return dst
+}
+
+func jacobianSerial(dst *mat64.Dense, f func([]float64, []float64), x, origin []float64, formula Formula, step float64) {
+	m, n := dst.Dims()
+	xcopy := make([]float64, n)
+	y := make([]float64, m)
+	col := make([]float64, m)
+	for j := 0; j < n; j++ {
+		for i := range col {
+			col[i] = 0
+		}
+		for _, pt := range formula.Stencil {
+			if pt.Loc == 0 {
+				if origin == nil {
+					origin = make([]float64, m)
+					copy(xcopy, x)
+					f(origin, xcopy)
+				}
+				floats.AddScaled(col, pt.Coeff, origin)
+			} else {
+				copy(xcopy, x)
+				xcopy[j] += pt.Loc * step
+				f(y, xcopy)
+				floats.AddScaled(col, pt.Coeff, y)
+			}
+		}
+		dst.SetCol(j, col)
+	}
+	dst.Scale(1/step, dst)
+}
+
+func jacobianConcurrent(dst *mat64.Dense, f func([]float64, []float64), x, origin []float64, formula Formula, step float64, nWorkers int) {
+	m, n := dst.Dims()
+	for i := 0; i < m; i++ {
+		for j := 0; j < n; j++ {
+			dst.Set(i, j, 0)
+		}
+	}
+
+	var (
+		wg sync.WaitGroup
+		mu = make([]sync.Mutex, n) // Guard access to individual columns.
+	)
+	worker := func(jobs <-chan jacJob) {
+		defer wg.Done()
+		xcopy := make([]float64, n)
+		y := make([]float64, m)
+		yVec := mat64.NewVector(m, y)
+		for job := range jobs {
+			copy(xcopy, x)
+			xcopy[job.j] += job.pt.Loc * step
+			f(y, xcopy)
+			col := dst.ColView(job.j)
+			mu[job.j].Lock()
+			col.AddScaledVec(col, job.pt.Coeff, yVec)
+			mu[job.j].Unlock()
+		}
+	}
+	jobs := make(chan jacJob, nWorkers)
+	for i := 0; i < nWorkers; i++ {
+		wg.Add(1)
+		go worker(jobs)
+	}
+	workerOrigin := func() {
+		defer wg.Done()
+		origin = make([]float64, m)
+		xcopy := make([]float64, n)
+		copy(xcopy, x)
+		f(origin, xcopy)
+	}
+	var (
+		hasOrigin   bool
+		originCoeff float64
+	)
+	for _, pt := range formula.Stencil {
+		if pt.Loc == 0 {
+			hasOrigin = true
+			originCoeff = pt.Coeff
+			if origin == nil {
```

Comment:
Are you sure you tested with -race? This looks like a race condition to me. Multiple workerOrigin goroutines can be started, since making the origin non-nil does not happen before any part of the loop. Thus origin can be set several times, and it's not clear which will be the last. 

This is easily fixed by changing the loop to if !hasOrigin, and then setting it below launching the goroutine.